pennylane_snowflurry/utility/debug.py

```diff
@@ -0,0 +1,121 @@
+import pennylane as qml
+import numpy as np
+import pennylane_snowflurry.processing.custom_gates as custom
+import random
+
+
+
+def are_matrices_equivalent(matrix1, matrix2, tolerance=1e-9):
+    """
+    Checks if two matrices are equal up to a complex multiplicative factor.
+
+    Args:
+        matrix1 (ndarray): First matrix.
+        matrix2 (ndarray): Second matrix.
+        tolerance (float): Numerical tolerance for comparison.
+
+    Returns:
+        bool: True if the matrices are equal up to a complex factor, False otherwise.
+    """
+    
+    det = np.linalg.det(matrix2)
+    
+    if abs(det) < tolerance:
+        raise ValueError("cannot invert matrix. cannot check equivalence")
+    
+    if matrix1.shape != matrix2.shape:
+        return False
+
+    matrix2_inv = np.linalg.inv(matrix2)
+    id = np.round(matrix1 @ matrix2_inv, 4)
+    value = id[0][0]
+    
+    # check if matrix is diagonal
+    for i in range(id.shape[0]):
+        for j in range(id.shape[1]):
+            # if non-diagonal position is > 0, not equivalent
+            if i != j and abs(id[i][j]) > tolerance:
+                return False
+            # if diagonal positions are not equal, not equivalent
+            if i == j and abs(id[i][j] - value) > tolerance:
+                return False
+    return True
+
+
+def add_noise(tape : QuantumTape, t = 0.005):
+    """
+    adds noise for each gate present in a quantum tape
+
+    Args:
+        tape (QuantumTape): the tape you want to noisify
+        t (float, optional): the noise quantity from 0 to 1. Defaults to 0.005.
+    """
+    new_operations = []
+    for op in tape.operations:
+        new_operations += [op] + [qml.DepolarizingChannel(random.random() * t, w) for w in op.wires]
+    
+    new_tape = type(tape)(ops=new_operations, measurements=tape.measurements, shots=tape.shots)
+    return [new_tape], lambda results : results[0]
+
+def to_qasm(tape : QuantumTape) -> str:
+    """
+    turns a quantum tape into a qasm string
+    """
+    eq = {
+        "PauliX" : "x", "PauliY" : "y", "PauliZ" : "z", "Identity" : "id",
+        "RX" : "rx", "RY" : "ry", "RZ" : "rz", "PhaseShift" : "p", "Hadamard" : "h",
+        "S" : "s", "Adjoint(S)" : "sdg", "SX" : "sx", "Adjoint(SX)" : "sxdg", "T" : "t", "Adjoint(T)" : "tdg", 
+        "CNOT" : "cx", "CY" : "cy", "CZ" : "cz", "SWAP" : "swap",
+        "Z90" : "s", "ZM90" : "sdg", "X90" : "sx", "XM90" : "sxdg", "Y90" : "ry(pi/2)", "YM90" : "ry(3*pi/2)",
+        "TDagger" : "tdg", "CRY" : "cry"
+    }
+    total_string = ""
+    for op in tape.operations:
+        string = eq[op.name]
+        if len(op.parameters) > 0:            
+            string += "(" + str(op.parameters[0]) + ")" 
+        string += " "
+        string += ", ".join([f"q[{w}]" for w in op.wires])
+        string += ";"
+        total_string += string + "\n"
+    return total_string
+
+def arbitrary_circuit(tape : QuantumTape, measurement = qml.counts):
```

Comment:
not sure if this works all the time. Might want to remove it since it's not used